examples/cluster/plot_cluster_comparison.py

```diff
@@ -3,116 +3,174 @@
+clustering. For this example, the null dataset uses the
+same parameters as the dataset in the row above it, which
+represents a mismatch in the parameter values and the
+data structure.
+
+While these examples give some intuition about the
+algorithms, this intuition might not apply to very high
+dimensional data.
 """
 print(__doc__)
 
 import time
+import warnings
 
 import numpy as np
 import matplotlib.pyplot as plt
 
-from sklearn import cluster, datasets
+from sklearn import cluster, datasets, mixture
 from sklearn.neighbors import kneighbors_graph
 from sklearn.preprocessing import StandardScaler
+from itertools import cycle, islice
 
 np.random.seed(0)
 
+# ============
 # Generate datasets. We choose the size big enough to see the scalability
 # of the algorithms, but not too big to avoid too long running times
+# ============
 n_samples = 1500
 noisy_circles = datasets.make_circles(n_samples=n_samples, factor=.5,
                                       noise=.05)
 noisy_moons = datasets.make_moons(n_samples=n_samples, noise=.05)
 blobs = datasets.make_blobs(n_samples=n_samples, random_state=8)
 no_structure = np.random.rand(n_samples, 2), None
 
-colors = np.array([x for x in 'bgrcmykbgrcmykbgrcmykbgrcmyk'])
-colors = np.hstack([colors] * 20)
-
-clustering_names = [
-    'MiniBatchKMeans', 'AffinityPropagation', 'MeanShift',
-    'SpectralClustering', 'Ward', 'AgglomerativeClustering',
-    'DBSCAN', 'Birch']
-
-plt.figure(figsize=(len(clustering_names) * 2 + 3, 9.5))
+# Anisotropicly distributed data
+random_state = 170
+X, y = datasets.make_blobs(n_samples=n_samples, random_state=random_state)
+transformation = [[0.6, -0.6], [-0.4, 0.8]]
+X_aniso = np.dot(X, transformation)
+aniso = (X_aniso, y)
+
+# blobs with varied variances
+varied = datasets.make_blobs(n_samples=n_samples,
+                             cluster_std=[1.0, 2.5, 0.5],
+                             random_state=random_state)
+
+# ============
+# Set up cluster parameters
+# ============
+plt.figure(figsize=(9 * 2 + 3, 12.5))
 plt.subplots_adjust(left=.02, right=.98, bottom=.001, top=.96, wspace=.05,
                     hspace=.01)
 
 plot_num = 1
 
-datasets = [noisy_circles, noisy_moons, blobs, no_structure]
-for i_dataset, dataset in enumerate(datasets):
+default_base = {'quantile': .3,
+                'eps': .3,
+                'damping': .9,
+                'preference': -200,
+                'n_neighbors': 10,
+                'n_clusters': 3}
+
+datasets = [
+    (noisy_circles, {'damping': .77, 'preference': -240,
+                     'quantile': .2, 'n_clusters': 2}),
+    (noisy_moons, {'damping': .75, 'preference': -220, 'n_clusters': 2}),
+    (varied, {'eps': .18, 'n_neighbors': 2}),
+    (aniso, {'eps': .15, 'n_neighbors': 2}),
+    (blobs, {}),
+    (no_structure, {})]
+
+for i_dataset, (dataset, params) in enumerate(datasets):
+    # update parameters with dataset-specific values
+    defaults = default_base.copy()
+    defaults.update(params)
```

Comment:
In thinking about this, I can see how we might move the estimators out of the loop and then update the parameters with `set_params`, but I don't see how we get rid of the defaults variable without replacing it with something more verbose.  

Since not all keywords need to be changed for each dataset, I would have to have another bit of code that checks to see if a keyword that we need is in `params` or not for that dataset.  If it isn't, I'd have to revert back to the `default_base` values.  I think updating the keyword dict `defaults`, as we have, is the simplest way to have parameters that get changed by dataset-estimator.  

The other option is to have the `params` have both the `default_base` and dataset-specific values enumerated for each dataset, but with that it is much harder to see where the dataset parameter excursions are.

Let me know if I am missing your point.